Convert jest testing over to TypeScript

TheCleric opened the pull request:
This PR will allow jest to run directly against the TypeScript files so we don't have to `npm run build` before testing and now `npm run test:watch" actually works!

ronjouch commented on app/tsconfig.json:
```diff
@@ -1,21 +1,24 @@
+    "sourceMap": true,
+    // https://stackoverflow.com/questions/51716406/typescript-tsconfig-settings-for-node-js-10
+    // and 'dom' to tell tsc it's okay to use the URL object (which is in Node >= 7)
+    "target": "es2019",
+    "lib": [
+      "es2020",
+      "dom"
+    ]
```
Careful with this, this will break users that for some reason keep using old versions of Electron (which come with older Node/v8).

TheCleric replied:
> Careful with this, this will break users that for some reason keep using old versions of Electron (which come with older Node/v8).

Ahh, was just trying to align it with the one is the actual project. Do we want to support that use case? If so, I can revert.

ronjouch replied:
> Ahh, was just trying to align it with the one is the actual project. Do we want to support that use case? If so, I can revert.

@TheCleric got it. This tsconfig.json should (like the main one) be bump as rarely as possible, to support running on old runtimes. Here in `app`, "runtime" means "the chosen Electron", whereas in `.`, "runtime" means "the node.js version of the user"

So, there's no alignment to aim at necessarily, we can/should bump it if an `app` dep needs it, but before this happens it's actually preferable to let it be, for users to keep being able to use old Electrons.

Adding a comment to clarify this.

ronjouch commented on package.json:
```diff
@@ -76,25 +76,28 @@
     "eslint-plugin-prettier": "^3.3.1",
-    "jest": "^26.6.3",
+    "jest": "^27.0.1",
+    "node-notifier": "^9.0.0",
     "prettier": "^2.2.1",
     "rimraf": "^3.0.2",
+    "ts-jest": "^27.0.0",
```
bleargh

ronjouch reviewed (commented):
> This PR will allow jest to run directly against the TypeScript files so we don't have to npm run build before testing and now `npm run test:watch" actually works!

`npm run test:watch` already works now already if you follow [HACKING.md / Tests](https://github.com/nativefier/nativefier/blob/master/HACKING.md#tests), which says:

> For a good live experience, open two terminal panes/tabs running code/tests watchers:
> 
> 1. Run a TSC watcher: `npm run build:watch`
> 2. Run a Jest unit tests watcher: `npm run test:watch`

This is what I'm doing in all my TS projects, work and personal, and I prefer it to what you're proposing in this PR, for the following reasons:

1. Most importantly: compile errors are neatly in one terminal tab/pane, while test results are in another. I find mixing both things to be confusing. Letting `tsc -w` do the source watching & compiling, and jest watch & test the output of `tsc -w` keeps responsibilities separated, with the UI intended by `tsc -w`, and user-visible.
2. The test runner doesn't need to know about TS.
3. Less deps, less conf.

Have you tried the workflow I recommend above? Does it not work? If it works, do you still have reasons to prefer what this PR suggests?

Note: if it works but you just weren't aware, what about pre/post-fixing `test:watch` with a little `echo 'Be sure to run the TS watcher (in another terminal pane/tab) to get a working test watcher! npm run build:watch'

TheCleric commented:
> This is what I'm doing in all my TS projects, work and personal, and I prefer it to what you're proposing in this PR, for the following reasons:
> 
> 1. Most importantly: compile errors are neatly in one terminal tab/pane, while test results are in another. I find mixing both things to be confusing. Letting `tsc -w` do the source watching & compiling, and jest watch & test the output of `tsc -w` keeps responsibilities separated, with the UI intended by `tsc -w`, and user-visible.

To me this isn't an issue, but I could see why someone would want them separate.

> 2. The test runner doesn't need to know about TS.

> 3. Less deps, less conf.

True, but these being dev-dependencies they don't concern me as much.

> Have you tried the workflow I recommend above? Does it not work? If it works, do you still have reasons to prefer what this PR suggests?

I do in the sense that
1. It's less that needs to be actively done (one command vs. two, which I frequently forget to do).
2. It feels more like the process that I do in other projects and language. Code some changes, run my tests, repeat until they're green. I think this dovetails with my next point which is:
3. To me this flow makes the JavaScript too prominent in the process. Honestly I don't care about the JavaScript. From my perspective, it's not much different than a compiled binary. It exists only to be a format that can be distributed to end users. The only thing I care about is writing the TypeScript, making sure it does what I want it to do (in unit tests and actual testing), and shipping it. The JavaScript is just a middle man that I avoid as much as possible. I'm sure that there's a chance I'm being overly simplistic or dismissive, so feel free to correct me here.

In the long run reason #3 above is 90% of why i did this, so if there's a difference of opinion here we can discuss that. 😄

ronjouch commented:
> 1. It's less that needs to be actively done (one command vs. two...

C'mon. We're working hours on feature/fixes, "one command vs. two" is peanuts 🙂.

> ... which I frequently forget to do).

My suggestion to prepend `test:watch` with something like `echo 'Remember to run npm run build:watch for the test watcher to run!' && ` addresses this.

> It feels more like the process that I do in other projects and language. Code some changes, run my tests, repeat until they're green

I don't understand this point. You can do this manual workflow (just `npm test`), and you can also run a watcher, by running the two `:watch` tasks.

> To me this flow makes the JavaScript too prominent in the process.

I don't see having build watcher separate from test watcher as making JS prominent. If anything, it's better separation of concerns, and more importantly, it's about using tools as close as possible to their default conf and without plugins. Tooling is annoying; precisely-finnickily-configured tooling is ***hell*** to maintain & keep up with & keep working over time. So, I want us to stay close to the ideal default config, and the added value you're exposing here doesn't feel strong enough to justify sacrificing tooling simplicity.

> Honestly I don't care about the JavaScript. From my perspective, it's not much different than a compiled binary. It exists only to be a format that can be distributed to end users. The only thing I care about is writing the TypeScript, making sure it does what I want it to do (in unit tests and actual testing), and shipping it. The JavaScript is just a middle man that I avoid as much as possible. I'm sure that there's a chance I'm being overly simplistic or dismissive, so feel free to correct me here.

Be veeeeeery careful with this. Not only you should care about the JS, because it's what our users run (and not TS), but you should care about the JS-after-being-build-then-npm-packed-then-npm-installed. See https://github.com/nativefier/nativefier/commit/cde5c1e13bdc2739604cab04bac64eae0d719ed1 for an example of this.

With all this said,

- I'm closing this
    - As usual: nothing personal, I sincerely appreciate your work, I'm taking this decision because I care, because I don't feel like I'm harming at all your productivity with it, and because it's what feels right for long-term maintainability.
- Pushing a commit with the aforementioned `echo '...' && ` to remind users of `test:watch` to run `build:watch` too.
    - Ideally we'd have do something more automatic (e.g. automating running both in separate tabs or at least exiting if the build watcher isn't running), and I actually have something like this at work, but it's not cross-platform because it uses `pwdx` (and obviously I don't want to add a dep to reach the same goal). An `echo` is good enough.
- The `app` tsconfig changes are I think undesired, until we actually need to guarantee a more recent Electron e.g. when we bump an `app` dependency requiring a recent Electron

TheCleric commented:
> - I'm closing this
> 
>     - As usual: nothing personal, I sincerely appreciate your work, I'm taking this decision because I care, because I don't feel like I'm harming at all your comfort/productivity with it, and because it's what feels right for long-term maintainability.

👍

I still disagree, but know that you are not doing this selfishly, but only because you feel it's best for the project. And I've been wrong before, so it's always a risk here as well. 😆

TheCleric commented:
@ronjouch I promise I'm not trying to rehash this, but am trying to work through our recommended flow, however I'm seeing an issue.

For the past while, I had found it easier to use `npm run build && npm run test` but was trying to get back into the habit you suggested above, which would be more efficient. I started two terminals, one for a build:watch and one for a test:watch, and it doesn't quite behave the way I think it would. For example, if I have those running and do something simple like changing line 137 of current master version of `src/options/optionsMain.ts` from:
```typescript
    const requestedMajorVersion = parseInt(requestedVersion.split('.')[0], 10);
```
to:
```typescript
    const requestedMajorVersion = parseInt(requestedVersion.split('.')[0], 11);
```
triggers tsc to rebuild, but no tests appear to run.

If I make a minor change in a test, such as changing line 16 in `src/options/optionsMain.ts` from:
```typescript
    targetUrl: 'https://example.com/',
```
to:
```typescript
    targetUrl: 'https://example2.com/',
```
then tsc doesn't run and no jest tests run.

Am I doing something wrong?

ronjouch commented:
@TheCleric 

> I promise I'm not trying to rehash this, but ...

Absolutely rehash this if it's broken for you 👍.

> am trying to work through our recommended flow, however I'm seeing an issue.

I read your thing, tried locally, noticed something wrong, pushed it to `master`, and was about to write here "fixed it, pls retry!", but re-reading your above comment, I now understand that for you, `test:watch` doesn't run *enough*.

Funnily, the issue I noticed locally (and fixed in 9c5dba7f07e03d678f835c1d2ef703a3b4ebcaf2) is that `test:watch` used to run *too much* for me, producing the annoying behavior that you had to save twice for your changed code to go through Jest, because your first save would immediately fire Jest, but on not-recompiled-yet JS (see comments in commit). So, I'd be surprised my commit fixes your problem (where neither tsc nor jest run), but it was a problem needing fixing, so can you nevertheless pull and retry, just to be sure.

With this fix, here is my experience on my Linux box. Can try on a window box if you want me to (and I know you have a Windows box/vm too, can you try there too?). Is the screencast of any help? By the way, it'll be a good illustration of how this workflow is pleasant; if we succeed at fixing this, I'll embed it in our dev guide to show future contributors what to expect.

https://user-images.githubusercontent.com/522085/120407694-abdf3f00-c31b-11eb-9ab5-a531a929adb9.mp4

Then, if the screencast doesn't help, something that for sure needs addressing on your side is that _"If I make a minor change in a test [...] then tsc doesn't run"_ . This means that before we even talk about Jest, there's the problem that tsc's watcher doesn't work for you, and that's not good (even if, somehow, jest-ts's watcher does work on your system; maybe jest-ts use a different lib / node-api? I dunno). Can you dig why `test:watch` doesn't work for you (googling, trying tsc's verbose/debug mode)? Is `tsc -w` broken for you just in Nativefier, or can you repro the issue in a minimal TS project? Maybe some OS configuration?

ronjouch commented:
> Can try on a window box if you want me to (and I know you have a Windows box/vm too, can you try there too?)

@TheCleric I just tested on a Windows 10 machine, `build:watch` and `test:watch` work fine too.

ronjouch commented:
@TheCleric also! Does `build:watch` work if you run it alone? If yes, could it be that `test:watch` simply exhausts all your OS file watch handles? Then, does adding `"/node_modules/",` to `watchPathIgnorePatterns` solve your issue?

TheCleric commented:
Okay so after watching the screencast, I think I see what may have been going wrong, and what we may be able to do to fix it. I see you go immediately and type in `a` to run all tests. I was expecting it to find the tests to run. After doing something similar to what you do here, things are running on save (though like you it seems to have some "run twice" issues). As well it seems that `build:watch` only looks for changes, but old builds may be cached between changing branches. For example a test I wrote in one branch was failing on master as it hadn't been deleted yet.

So here are my suggestions (along with your fix you spoke of earlier):

1. Change `build:watch` to run `npm run build && ...` so that old builds are cleaned out and you are starting out accurately.
2. Figure out a way for `test:watch` to actually watch for changes. I tried changing `--watch` to `--watchAll` but that produces a lot of output around code coverage that I wasn't looking for.

What do you think?

ronjouch commented:
@TheCleric all good feedback; addressed in 6b09d14. How does it feel now?

TheCleric commented:
> @TheCleric all good feedback; addressed in [6b09d14](https://github.com/nativefier/nativefier/commit/6b09d1467f907429e42633dcca8db6ff71f4283d). How does it feel now?
> 
> Also, I brought the upgrade to Jest 27 (with the necessary jasmine.spyOn → jest.spyOn fix) that was in this PR and I had missed as I closed it. Thanks 🙏

Much better. Especially since I just saw `npm run watch` that integrates the two, I think we've come to a nice middle ground. Thanks!
